prog: TestSerializeDeserializeRandom failed

Observed on revision cfebc8873b2f707ef8ed0bb2d8abb0280fede23a

```
--- FAIL: TestSerializeDeserializeRandom (8.12s)
    testutil.go:35: seed=1750120359522473471
    --- FAIL: TestSerializeDeserializeRandom/test/64 (0.01s)
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
            test$parent_conditions(&(0x7f00000001c0)={0x4, @without_flag1=0x95, {0x6, @value=0x2}})
            test$parent_conditions(&(0x7f0000000200)={0xff, @without_flag1=0x6, {0x1, @value=0x7}})
            test$parent_conditions(&(0x7f0000000240)={0xec, @without_flag1=0x7, {0x4, @value=0x7f}})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
            test$parent_conditions(&(0x7f00000001c0)={0x4, @without_flag1=0x95, {0x6, @value=0x2}})
            test$parent_conditions(&(0x7f0000000200)={0xff, @without_flag1=0x6, {0x1, @value=0x7}})
            test$parent_conditions(&(0x7f0000000240)={0xec, @without_flag1=0x7, {0x4, @value=0x7f}})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
            test$parent_conditions(&(0x7f00000001c0)={0x4, @without_flag1=0x95, {0x6, @value=0x2}})
            test$parent_conditions(&(0x7f0000000200)={0xff, @without_flag1=0x6, {0x1, @value=0x7}})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
            test$parent_conditions(&(0x7f00000001c0)={0x4, @without_flag1=0x95, {0x6, @value=0x2}})
            test$parent_conditions(&(0x7f0000000200)={0xff, @without_flag1=0x6, {0x1, @value=0x7}})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
            test$parent_conditions(&(0x7f00000001c0)={0x4, @without_flag1=0x95, {0x6, @value=0x2}})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
            test$parent_conditions(&(0x7f00000001c0)={0x4, @without_flag1=0x95, {0x6, @value=0x2}})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
            test$parent_conditions(&(0x7f0000000180)={0xc912, @without_flag1=0x5, {0x7fffffffffffffff}})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
            test$parent_conditions(&(0x7f0000000140)={0x84, @without_flag1=0x4, {0x2, @value=0x1fc0000}})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
            test$parent_conditions(&(0x7f0000000100)={0x0, @without_flag1=0x2, {0x6}})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f0000000080)={0xc87c, @without_flag1=0x10001, {0x100000001, @value=0xffffffff}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f0000000040)={0x3, @with_flag1=0x7, {0x7}})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f00000000c0)={0x3})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f0000000000)={0xa3, @without_flag1=0x93b})
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f00000000c0)={0x3})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
        encoding_test.go:449: PROG0:
            test$parent_conditions(&(0x7f00000000c0)={0x3})
        encoding_test.go:450: PROG1:
            test$parent_conditions(&(0x7f00000000c0)={0x3, @with_flag1})
        encoding_test.go:427: decoded0: {Calls:[{Meta:0xc000624700 Props:{FailNth:0 Async:false Rerun:0} Index:18446744073709551615 Args:[{Size:8 Format:0 Value:35184372089024 BitfieldOffset:0 BitfieldLength:0 PidStride:0}] Copyin:[{Addr:35184372089024 Arg:{Size:4 Format:0 Value:3 BitfieldOffset:0 BitfieldLength:0 PidStride:0}} {Addr:35184372089028 Arg:{Size:8 Format:0 Value:0 BitfieldOffset:0 BitfieldLength:0 PidStride:0}} {Addr:35184372089036 Arg:{Size:8 Format:0 Value:0 BitfieldOffset:0 BitfieldLength:0 PidStride:0}}] Copyout:[]}] Vars:[]}
        encoding_test.go:428: decoded1: {Calls:[{Meta:0xc000624700 Props:{FailNth:0 Async:false Rerun:0} Index:18446744073709551615 Args:[{Size:8 Format:0 Value:35184372089024 BitfieldOffset:0 BitfieldLength:0 PidStride:0}] Copyin:[{Addr:35184372089024 Arg:{Size:4 Format:0 Value:3 BitfieldOffset:0 BitfieldLength:0 PidStride:0}} {Addr:35184372089028 Arg:{Size:4 Format:0 Value:0 BitfieldOffset:0 BitfieldLength:0 PidStride:0}} {Addr:35184372089032 Arg:{Size:8 Format:0 Value:0 BitfieldOffset:0 BitfieldLength:0 PidStride:0}}] Copyout:[]}] Vars:[]}
        encoding_test.go:429: was: "\x02\x03\x80\x03\x00\b\x06\x03\x88\x03\x00\x10\x00\x03\x98\x03\x00\x10\x00\xec\x02\x01\x02\x04\x80\x03\x01"
            got: "\x02\x03\x80\x03\x00\b\x06\x03\x88\x03\x00\b\x00\x03\x90\x03\x00\x10\x00\xec\x02\x01\x02\x04\x80\x03\x01"
            program:
            test$parent_conditions(&(0x7f00000000c0)={0x3})
FAIL
FAIL	github.com/google/syzkaller/prog	35.486s
```

--- a/prog/expr.go
+++ b/prog/expr.go
@@ -205,8 +205,11 @@ func (c *Call) setDefaultConditions(target *Target, transientOnly bool) bool {
 				if transientOnly && !unionArg.transient {
 					return
 				}
-				// If several union options match, take the first one.
 				idx := okIndices[0]
+				if defIdx, ok := unionType.defaultField(); ok {
+					// If there's a default value available, use it.
+					idx = defIdx
+				}
 				field := unionType.Fields[idx]
 				replace[unionArg] = MakeUnionArg(unionType,
 					unionArg.Dir(),

--- a/prog/expr_test.go
+++ b/prog/expr_test.go
@@ -10,6 +10,7 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
 func TestGenerateConditionalFields(t *testing.T) {
@@ -308,4 +309,17 @@ func TestNestedConditionalCall(t *testing.T) {
 			}
 		}
 	}
+}
+
+func TestDefaultConditionalSerialize(t *testing.T) {
+	// Serialize() omits default-valued fields for a more compact representation,
+	// but that shouldn't mess with the selected option (see #6105).
+	const rawProg = "test$parent_conditions(&(0x7f0000000000)={0xa3})\n"
+	target := initTargetTest(t, "test", "64")
+	prog, err := target.Deserialize([]byte(rawProg), NonStrict)
+	require.NoError(t, err)
+	serialized := prog.Serialize()
+	prog2, err := target.Deserialize(serialized, NonStrict)
+	require.NoError(t, err)
+	assert.Equal(t, rawProg, string(prog2.Serialize()))
 }

--- a/prog/types.go
+++ b/prog/types.go
@@ -815,31 +815,43 @@ func (t *UnionType) String() string {
 }
 
 func (t *UnionType) DefaultArg(dir Dir) Arg {
-	idx := t.defaultField()
+	idx, _ := t.defaultField()
 	f := t.Fields[idx]
 	arg := MakeUnionArg(t, dir, f.DefaultArg(f.Dir(dir)), idx)
 	arg.transient = t.isConditional()
 	return arg
 }
 
-func (t *UnionType) defaultField() int {
-	// If it's a conditional union, the last field will be the default value.
+func (t *UnionType) defaultField() (int, bool) {
+	// If it's a conditional union, the last field is usually a safe choice for the default value as
+	// it must have no condition.
+	// Auto-generated wrappers for conditional fields are an exception since both fields will have
+	// conditions, and, moreover, these conditions will be mutually exclusive.
 	if t.isConditional() {
-		return len(t.Fields) - 1
+		if t.Fields[len(t.Fields)-1].Condition != nil {
+			// There's no correct default index.
+			return 0, false
+		}
+		return len(t.Fields) - 1, true
 	}
 	// Otherwise, just take the first.
-	return 0
+	return 0, true
 }
 
 func (t *UnionType) isConditional() bool {
-	// In pkg/compiler, we ensure that either none of the fields have conditions,
-	// or all except the last one.
+	// Either all fields will have a conditions, or all except the last one, or none.
+	// So checking for the first one is always enough.
 	return t.Fields[0].Condition != nil
 }
 
 func (t *UnionType) isDefaultArg(arg Arg) bool {
 	a := arg.(*UnionArg)
-	return a.Index == t.defaultField() && isDefault(a.Option)
+	defIdx, ok := t.defaultField()
+	if !ok {
+		// Any value is the only possible option.
+		return isDefault(a.Option)
+	}
+	return a.Index == defIdx && isDefault(a.Option)
 }
 
 type ConstValue struct {

--- a/prog/validation.go
+++ b/prog/validation.go
@@ -261,6 +261,10 @@ func (arg *UnionArg) validate(ctx *validCtx, dir Dir) error {
 	if arg.Index < 0 || arg.Index >= len(typ.Fields) {
 		return fmt.Errorf("union arg %v has bad index %v/%v", arg, arg.Index, len(typ.Fields))
 	}
+	if arg.transient && !ctx.opts.ignoreTransient {
+		// The union must have been patched via Call.setDefaultConditions.
+		return fmt.Errorf("union arg %v is transient (incomplete)", arg)
+	}
 	opt := typ.Fields[arg.Index]
 	return ctx.validateArg(arg.Option, opt.Type, opt.Dir(dir))
 }

